File: src/simulation/culture/thinking/meaning/MemePool.java
package simulation.culture.thinking.meaning;


import java.util.*;

public class MemePool {
    private Map<String, Meme> memes;

    private MemePool(Collection<Meme> memes) {
        this.memes = new HashMap<>();
        memes.forEach(this::add);
    }

    MemePool() {
        this(new ArrayList<>());
    }

    public boolean add(Meme meme) {
        if (memes.containsKey(meme.toString())) {
            return false;
        }
        memes.put(meme.toString(), meme.copy());
        return true;
    }

    public void addAll(Collection<Meme> memes) {
        memes.forEach(this::add);
    }

    public void addAll(MemePool memePool) {
        memePool.memes.values().forEach(this::add);
    }

    public List<Meme> getMemes() {
        return new ArrayList<>(memes.values());
    }

    public Meme getMeme(String name) {
        return memes.get(name.toLowerCase());
    }

    public Meme getMemeCopy(String name) {
        Meme meme = getMeme(name);
        return meme == null ? null : meme.copy();
    }

    public boolean isEmpty() {
        return memes.isEmpty();
    }

    public boolean strengthenMeme(Meme meme) {
        return strengthenMeme(meme, 1);
    }

    public boolean strengthenMeme(Meme meme, int delta) {
        Meme existing = memes.get(meme.toString());
        if (existing != null) {
            existing.increaseImportance(delta);
            return true;
        }
        return false;
    }
}

File: src/simulation/culture/thinking/meaning/MemePool.kt
package simulation.culture.thinking.meaning

import java.util.*


open class MemePool private constructor(memes: Collection<Meme>) {
    internal constructor() : this(ArrayList<Meme>())

    protected val memes = mutableMapOf<String, Meme>()

    init {
        addAll(memes)
    }

    open val isEmpty: Boolean
        get() = memes.isEmpty()

    fun add(meme: Meme): Boolean {
        if (memes.containsKey(meme.toString()))
            return false

        memes[meme.toString()] = meme.copy()

        return true
    }

    fun addAll(memes: Collection<Meme>) = memes.forEach { add(it) }

    fun addAll(pool: MemePool) = addAll(pool.memes.values)

    open val all get() = memes.values.toList()

    open fun getMeme(name: String) = memes[name.toLowerCase()]//TODO wat?

    open fun getMemeCopy(name: String) = getMeme(name)?.copy()

    open fun strengthenMeme(meme: Meme) = strengthenMeme(meme, 1)

    open fun strengthenMeme(meme: Meme, delta: Int) = memes[meme.toString()]
            ?.let {
                it.increaseImportance(delta)
                return true
            } ?: false
}
